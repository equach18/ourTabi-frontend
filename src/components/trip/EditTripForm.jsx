import { useState, useEffect } from "react";
import { useNavigate, useParams } from "react-router-dom";
import Alert from "../common/Alert";

function EditTripForm({ editTrip, trips }) {
  const { tripId } = useParams();
  const navigate = useNavigate();
  const trip = trips.find((t) => t.id === Number(tripId));

  useEffect(() => {
    if (!trip) {
      navigate("/not-found");
    }
  }, [trip, navigate]);

  // initialize form state with formatted date values
  const [formData, setFormData] = useState({
    title: trip.title || "",
    destination: trip.destination || "",
    radius: trip.radius || 10,
    startDate: trip.startDate
      ? new Date(trip.startDate).toISOString().split("T")[0]
      : "",
    endDate: trip.endDate
      ? new Date(trip.endDate).toISOString().split("T")[0]
      : "",
    isPrivate: trip.isPrivate || false,
  });

  const [isSubmitting, setIsSubmitting] = useState(false);
  const [formErrors, setFormErrors] = useState([]);

  function handleChange(evt) {
    const { name, value, type, checked } = evt.target;
    setFormData((data) => ({
      ...data,
      [name]: type === "checkbox" ? checked : value,
    }));
  }

  async function handleSubmit(evt) {
    evt.preventDefault();
    setIsSubmitting(true);
    setFormErrors([]);

    try {
      await editTrip(trip.id, formData);
      navigate(`/trips/${trip.id}`);
    } catch (err) {
      setFormErrors(["Error updating trip. Please try again."]);
    } finally {
      setIsSubmitting(false);
    }
  }

  return (
    <div className="h-[calc(100vh-4rem)] flex items-center justify-center px-4 py-10">
      <div className="w-full max-w-xl bg-white p-8 rounded-xl shadow-lg">
        <h1 className="text-3xl font-bold text-gray-600 mb-4">Edit Trip</h1>

        <form onSubmit={handleSubmit} className="space-y-3">
          {/* title */}
          <div>
            <label className="block text-zinc-700 font-medium mb-1" htmlFor="title">
              Trip Title
            </label>
            <input
              type="text"
              name="title"
              id="title"
              value={formData.title}
              onChange={handleChange}
              className="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
              required
            />
          </div>

          {/* destination */}
          <div>
            <label className="block text-zinc-700 font-medium mb-1" htmlFor="destination">
              Destination
            </label>
            <input
              type="text"
              name="destination"
              id="destination"
              value={formData.destination}
              onChange={handleChange}
              className="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
              required
            />
          </div>

          {/* radius */}
          <div>
            <label className="block text-zinc-700 font-medium mb-1" htmlFor="radius">
              Radius (miles/km)
            </label>
            <input
              type="number"
              name="radius"
              id="radius"
              value={formData.radius}
              onChange={handleChange}
              className="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
              min="1"
              required
            />
          </div>

          {/* start date */}
          <div>
            <label className="block text-zinc-700 font-medium mb-1" htmlFor="startDate">
              Start Date
            </label>
            <input
              type="date"
              name="startDate"
              id="startDate"
              value={formData.startDate || ""}
              onChange={handleChange}
              className="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            />
          </div>

          {/* end date */}
          <div>
            <label className="block text-zinc-700 font-medium mb-1" htmlFor="endDate">
              End Date
            </label>
            <input
              type="date"
              name="endDate"
              id="endDate"
              value={formData.endDate || ""}
              onChange={handleChange}
              className="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
              required={!!formData.startDate}
              disabled={!formData.startDate}
            />
          </div>

          {/* privacy */}
          <div className="flex items-center space-x-2">
            <input
              type="checkbox"
              name="isPrivate"
              id="isPrivate"
              checked={formData.isPrivate}
              onChange={handleChange}
              className="w-4 h-4 text-emerald-600 border-gray-300 rounded focus:ring-emerald-500"
            />
            <label className="text-zinc-700" htmlFor="isPrivate">Private Trip</label>
          </div>

          {/* error alerts */}
          <Alert type="error" messages={formErrors} />

          <div className="flex gap-4 pt-2">
            <button
              type="submit"
              className={`w-full bg-blue-500 text-white py-2 rounded-lg hover:bg-blue-600 transition ${
                isSubmitting && "opacity-50 cursor-not-allowed"
              }`}
              disabled={isSubmitting}
            >
              {isSubmitting ? "Updating..." : "Save Changes"}
            </button>

            <button
              type="button"
              className="w-full bg-gray-400 text-white py-2 rounded-lg hover:bg-gray-500 transition"
              onClick={() => navigate(`/trips/${trip.id}`)}
            >
              Cancel
            </button>
          </div>
        </form>
      </div>
    </div>
  );
}

export default EditTripForm;
