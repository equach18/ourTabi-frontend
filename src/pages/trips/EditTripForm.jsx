import { useState, useEffect } from "react";
import { useNavigate, useParams } from "react-router-dom";

function EditTripForm({ editTrip, trips }) {
  const { tripId } = useParams();
  const navigate = useNavigate();
  const trip = trips.find((t) => t.id === Number(tripId));

  useEffect(() => {
    if (!trip) {
      navigate("/not-found");
    }
  }, [trip, navigate]);

  // Initialize form state with formatted date values
  const [formData, setFormData] = useState({
    title: trip.title || "",
    destination: trip.destination || "",
    radius: trip.radius || 10,
    startDate: trip.startDate
      ? new Date(trip.startDate).toISOString().split("T")[0]
      : "",
    endDate: trip.endDate
      ? new Date(trip.endDate).toISOString().split("T")[0]
      : "",
    isPrivate: trip.isPrivate || false,
  });

  const [isSubmitting, setIsSubmitting] = useState(false);
  const [formErrors, setFormErrors] = useState([]);

  /** Handles input changes */
  function handleChange(evt) {
    const { name, value, type, checked } = evt.target;
    setFormData((data) => ({
      ...data,
      [name]: type === "checkbox" ? checked : value,
    }));
  }

  /** Handles form submission */
  async function handleSubmit(evt) {
    evt.preventDefault();
    setIsSubmitting(true);
    setFormErrors([]);

    try {
      await editTrip(trip.id, formData);
      navigate(`/trips/${trip.id}`);
    } catch (err) {
      setFormErrors(["Error updating trip. Please try again."]);
    } finally {
      setIsSubmitting(false);
    }
  }

  return (
    <div className="max-w-lg mx-auto mt-10 p-6 bg-white shadow-lg rounded-lg">
      <h1 className="text-2xl font-bold text-blue-500">Edit Trip</h1>

      <form onSubmit={handleSubmit} className="mt-4">
        {/* Title */}
        <div className="mb-4">
          <label className="block text-gray-700">Trip Title</label>
          <input
            type="text"
            name="title"
            value={formData.title}
            onChange={handleChange}
            className="w-full p-2 border rounded-lg"
            required
          />
        </div>

        {/* Destination */}
        <div className="mb-4">
          <label className="block text-gray-700">Destination</label>
          <input
            type="text"
            name="destination"
            value={formData.destination}
            onChange={handleChange}
            className="w-full p-2 border rounded-lg"
            required
          />
        </div>

        {/* Radius */}
        <div className="mb-4">
          <label className="block text-gray-700">Radius (miles/km)</label>
          <input
            type="number"
            name="radius"
            value={formData.radius}
            onChange={handleChange}
            className="w-full p-2 border rounded-lg"
            min="1"
            required
          />
        </div>

        {/* Start Date */}
        <div className="mb-4">
          <label className="block text-gray-700">Start Date</label>
          <input
            type="date"
            name="startDate"
            value={formData.startDate || ""}
            onChange={handleChange}
            className="w-full p-2 border rounded-lg"
          />
        </div>

        {/* End Date */}
        <div className="mb-4">
          <label className="block text-gray-700">End Date</label>
          <input
            type="date"
            name="endDate"
            value={formData.endDate || ""}
            onChange={handleChange}
            className="w-full p-2 border rounded-lg"
            required={!!formData.startDate} // Only required if startDate exists
            disabled={!formData.startDate} // Disable unless startDate is filled
          />
        </div>

        {/* Privacy Checkbox */}
        <div className="mb-4">
          <label className="inline-flex items-center">
            <input
              type="checkbox"
              name="isPrivate"
              checked={formData.isPrivate}
              onChange={handleChange}
              className="mr-2"
            />
            Private Trip
          </label>
        </div>

        {/* Error Messages */}
        {formErrors.length > 0 && (
          <div className="text-red-500">
            {formErrors.map((err, i) => (
              <p key={i}>{err}</p>
            ))}
          </div>
        )}

        {/* Submit & Cancel Buttons */}
        <div className="flex space-x-4">
          <button
            type="submit"
            className="bg-blue-500 text-white px-6 py-3 rounded-lg hover:bg-blue-600 w-full"
            disabled={isSubmitting}
          >
            {isSubmitting ? "Updating..." : "Save Changes"}
          </button>
          <button
            type="button"
            className="bg-gray-400 text-white px-6 py-3 rounded-lg w-full"
            onClick={() => navigate(`/trips/${trip.id}`)}
          >
            Cancel
          </button>
        </div>
      </form>
    </div>
  );
}

export default EditTripForm;
