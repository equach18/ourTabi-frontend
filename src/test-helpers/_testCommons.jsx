import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import UserContext from "../context/UserContext";
import TripContext from "../context/TripContext";

export const testUser1 = {
  id: 1,
  firstName: "Test1",
  lastName: "User1",
  username: "testUser1",
  profilePic: "",
};
export const testUser2 = {
  id: 2,
  firstName: "Test2",
  lastName: "User2",
  username: "testUser2",
  profilePic: "testUser2.png",
};
export const testUser3 = {
  id: 3,
  firstName: "Test2",
  lastName: "User2",
  username: "testUser3",
  profilePic: "testUser3.png",
};

export const defaultMembers = [
  {
    userId: testUser1.id,
    username: testUser1.username,
    profilePic: testUser1.profilePic,
  },
  {
    userId: testUser2.id,
    username: testUser2.username,
    profilePic: testUser2.profilePic,
  },
];

export const testActivity = {
  id: 1,
  name: "Kayaking",
  tripId: 53,
  category: "Other",
  location: "Chicago",
  description: "So much fun",
  scheduledTime: "2025-07-10T14:30:00Z",
};

export const testComment = {
  id: 1,
  text: "Test comment",
  createdAt: "2025-07-11T14:30:00Z",
  userId: 1,
};

export const testTrip = {
  id: 42,
  title: "Test trip",
  destination: "Chicago, IL",
  startDate: "2025-07-10",
  endDate: "2025-07-12",
  isPrivate: true,
};

export const testFriendsData = {
  friends: [testUser1 ],
  incomingRequests: [testUser2],
  sentRequests: [testUser3],
};

// Default mock context values
const defaultUserContext = { currentUser: testUser1 };
const defaultTripContext = {
  voteOnActivity: vi.fn(),
  votes: {},
  deleteComment: vi.fn(),
  addMember: vi.fn(),
  removeMember: vi.fn(),
  members: defaultMembers,
};

// create mock trip context
export function createMockTripContext(overrides = {}) {
  return {
    voteOnActivity: vi.fn(),
    votes: {},
    deleteComment: vi.fn(),
    members: [],
    addMember: vi.fn(),
    removeMember: vi.fn(),
    ...overrides,
  };
}

/**
 * Renders a component inside MemoryRouter
 */
export function renderWithRouter(ui) {
  return render(<MemoryRouter>{ui}</MemoryRouter>);
}

/**
 * Renders a component inside MemoryRouter + UserContext + TripContext
 */
export function renderWithAllProviders(
  ui,
  {
    userContext = defaultUserContext,
    tripContext = defaultTripContext,
    route = "/",
  } = {}
) {
  return render(
    <MemoryRouter initialEntries={[route]}>
      <UserContext.Provider value={userContext}>
        <TripContext.Provider value={tripContext}>{ui}</TripContext.Provider>
      </UserContext.Provider>
    </MemoryRouter>
  );
}

/**
 * Renders a component with just UserContext
 */
export function renderWithUserProvider(ui, { userContext } = {}) {
  return render(
    <UserContext.Provider value={userContext}>{ui}</UserContext.Provider>
  );
}
