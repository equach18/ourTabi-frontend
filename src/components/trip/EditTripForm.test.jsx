import { fireEvent, screen, waitFor } from "@testing-library/react";
import { Routes, Route } from "react-router-dom";
import EditTripForm from "./EditTripForm";
import { renderWithAllProviders } from "../../test-helpers/_testCommons";
const mockNavigate = vi.fn();

vi.mock("react-router-dom", async () => {
  const actual = await vi.importActual("react-router-dom");
  return {
    ...actual,
    useParams: () => ({ tripId: "24" }),
    useNavigate: () => mockNavigate,
  };
});

const testTrip = {
  id: 24,
  title: "Test Trip",
  destination: "Tokyo",
  radius: 25,
  startDate: "2025-07-01",
  endDate: "2025-07-05",
  isPrivate: true,
};

describe("EditTripForm", () => {
  const mockEditTrip = vi.fn();

  function renderForm() {
    return renderWithAllProviders(
      <Routes>
        <Route
          path="/trips/:tripId/edit"
          element={<EditTripForm editTrip={mockEditTrip} trips={[testTrip]} />}
        />
      </Routes>,
      { route: "/trips/24/edit" }
    );
  }

  beforeEach(() => {
    mockEditTrip.mockReset();
    mockNavigate.mockReset();
  });

  it("matches snapshot", () => {
    const { asFragment } = renderForm();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders form with initial trip values", () => {
    renderForm();
    expect(screen.getByDisplayValue("Test Trip")).toBeInTheDocument();
    expect(screen.getByDisplayValue("Tokyo")).toBeInTheDocument();
    expect(screen.getByDisplayValue("25")).toBeInTheDocument();
    expect(screen.getByDisplayValue("2025-07-01")).toBeInTheDocument();
    expect(screen.getByDisplayValue("2025-07-05")).toBeInTheDocument();
    expect(screen.getByLabelText(/private trip/i)).toBeChecked();
  });

    it("submits form and navigates on success", async () => {
      mockEditTrip.mockResolvedValue();

      renderForm();
      fireEvent.change(screen.getByLabelText(/destination/i), {
        target: { value: "Osaka" },
      });
      fireEvent.click(screen.getByRole("button", { name: /save changes/i }));

      await waitFor(() => {
        expect(mockEditTrip).toHaveBeenCalledWith(
          24,
          expect.objectContaining({ destination: "Osaka" })
        );
        expect(mockNavigate).toHaveBeenCalledWith("/trips/24");
      });
    });

    it("shows error msg if submission fail", async () => {
      mockEditTrip.mockRejectedValue(new Error("fail"));

      renderForm();
      fireEvent.click(screen.getByRole("button", { name: /save changes/i }));

      expect(await screen.findByText(/error updating trip/i)).toBeInTheDocument();
    });

    it("navigates to trip page when cancel btn is clicked", () => {
      renderForm();
      fireEvent.click(screen.getByRole("button", { name: /cancel/i }));
      expect(mockNavigate).toHaveBeenCalledWith("/trips/24");
    });

    it("disables end date if no start date", () => {
      const trip = { ...testTrip, startDate: null, endDate: null };
      renderWithAllProviders(
        <Routes>
          <Route
            path="/trips/:tripId/edit"
            element={<EditTripForm editTrip={mockEditTrip} trips={[trip]} />}
          />
        </Routes>,
        { route: "/trips/24/edit" }
      );

      const endDateInput = screen.getByLabelText(/end date/i);
      expect(endDateInput).toBeDisabled();
    });
});
