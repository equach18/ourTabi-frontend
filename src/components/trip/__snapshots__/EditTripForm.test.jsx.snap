// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditTripForm > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="h-[calc(100vh-4rem)] flex items-center justify-center px-4 py-10"
  >
    <div
      class="w-full max-w-xl bg-white p-8 rounded-xl shadow-lg"
    >
      <h1
        class="text-3xl font-bold text-gray-600 mb-4"
      >
        Edit Trip
      </h1>
      <form
        class="space-y-3"
      >
        <div>
          <label
            class="block text-zinc-700 font-medium mb-1"
            for="title"
          >
            Trip Title
          </label>
          <input
            class="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            id="title"
            name="title"
            required=""
            type="text"
            value="Test Trip"
          />
        </div>
        <div>
          <label
            class="block text-zinc-700 font-medium mb-1"
            for="destination"
          >
            Destination
          </label>
          <input
            class="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            id="destination"
            name="destination"
            required=""
            type="text"
            value="Tokyo"
          />
        </div>
        <div>
          <label
            class="block text-zinc-700 font-medium mb-1"
            for="radius"
          >
            Radius (miles/km)
          </label>
          <input
            class="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            id="radius"
            min="1"
            name="radius"
            required=""
            type="number"
            value="25"
          />
        </div>
        <div>
          <label
            class="block text-zinc-700 font-medium mb-1"
            for="startDate"
          >
            Start Date
          </label>
          <input
            class="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            id="startDate"
            name="startDate"
            type="date"
            value="2025-07-01"
          />
        </div>
        <div>
          <label
            class="block text-zinc-700 font-medium mb-1"
            for="endDate"
          >
            End Date
          </label>
          <input
            class="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            id="endDate"
            name="endDate"
            required=""
            type="date"
            value="2025-07-05"
          />
        </div>
        <div
          class="flex items-center space-x-2"
        >
          <input
            checked=""
            class="w-4 h-4 text-emerald-600 border-gray-300 rounded focus:ring-emerald-500"
            id="isPrivate"
            name="isPrivate"
            type="checkbox"
          />
          <label
            class="text-zinc-700"
            for="isPrivate"
          >
            Private Trip
          </label>
        </div>
        <div
          class="flex gap-4 pt-2"
        >
          <button
            class="w-full bg-blue-500 text-white py-2 rounded-lg hover:bg-blue-600 transition false"
            type="submit"
          >
            Save Changes
          </button>
          <button
            class="w-full bg-gray-400 text-white py-2 rounded-lg hover:bg-gray-500 transition"
            type="button"
          >
            Cancel
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
